### PR TITLE
Add option to use traditional threading in the UFS

### LINKED ISSUE #2277: Add traditional threading option for UFSWM
### What new functionality do you need?

ESMF-managed threading was added in #1371 for the ufs-weather-model.
A traditional threading option is also desired as some issues were reported while using ESMF-managed threading at higher resolutions (e.g. C1152).
An option to use either of those available threading options is desired.

### What are the requirements for the new functionality?

1. A configurable flag that allows one to choose between the two threading options when launching the ufs-weather-model forecast executable in the workflow.

### Acceptance Criteria

Successful execution of the model in the two threading modes.

### Suggest a solution (optional)

1. Add a flag in config.fcst/config.efcs/config.ufs or elsewhere where it makes sense to toggle the threading options
2. Choose the appropriate `ufs.configure` template.  Note that traditional threading templates will need to be linked in the global-workflow from the ufs-weather-model, and if templates are not available in the ufs-weather-model, they will need to be added there first.
3. update the calculations of PETS for the components and total taking into consideration the toggle in 1.
4. set `OMP_NUM_THREADS` when using traditional threading before calling the executable.

### LINKED ISSUE #2407: diag_table issue due to atparse.sh in parsing_namelists_FV3.sh
### What is wrong?

It seems there's a mistake in the atparse script. The issue lies in how the while loop is structured within the atparse function. Specifically, the loop termination condition [[ "$__after" == "$__text" ]] may not always work as expected, leading to premature termination of the loop and missing the last line. So the last line in the file of diag_table_aod is always missing in the finale diag_table, which results in the missing of sea salt AOD in the diagnostic output. 

### What should have happened?

In the atparse script, I've changed the while loop condition to while IFS= read -r __text || [ -n "$__text" ], which ensures that the loop continues until the input is fully consumed, even if the last line doesn't end with a newline character. It works well to solve the issue.

### What machines are impacted?

All or N/A

### Steps to reproduce

In the run directory, you can see that the last line of diag_table is always missing compared to the diag_table_template 

### Additional information

"gfs_phys",  "DU_AOD_550",      "du_aod550",     "fv3_history2d",  "all",  .false.,  "none",  2
"gfs_phys",  "SU_AOD_550",      "su_aod550",     "fv3_history2d",  "all",  .false.,  "none",  2
"gfs_phys",  "BC_AOD_550",      "bc_aod550",     "fv3_history2d",  "all",  .false.,  "none",  2
"gfs_phys",  "OC_AOD_550",      "oc_aod550",     "fv3_history2d",  "all",  .false.,  "none",  2
"gfs_phys",  "SS_AOD_550",      "ss_aod550",     "fv3_history2d",  "all",  .false.,  "none",  2

The last line is always missing in the diag_table

### Do you have a proposed solution?

_No response_

### LINKED ISSUE #2408: GFS Off-line post failed on WCOSS2
### What is wrong?

The GFS C768 case failed at off-line post on WCOSS2.

### What should have happened?

The job 'JGLOBAL_ATMOS_UPP' should have capacity to process high resolution model output with off-line post.

### What machines are impacted?

WCOSS2

### Steps to reproduce

Here are errors from my runtime log:
```
^[[38;5;196m2024-03-18 17:12:57,639 - ERROR    - upp         : FATAL ERROR: Error occurred during execution of mpiexec -l -n 120 -ppn 120 --cpu-bind depth --depth 1 /u/wen.meng/ptmp/gw_test/RUNDIR/upp.231494/upp.x^[[0m
Traceback (most recent call last):
  File "/u/wen.meng/noscrub/ncep_post/gw_v17/global-workflow/ush/python/pygfs/task/upp.py", line 239, in _call_executable
    exec_cmd()
  File "/u/wen.meng/noscrub/ncep_post/gw_v17/global-workflow/ush/python/wxflow/executable.py", line 230, in __call__
    raise ProcessError(f"Command exited with status {proc.returncode}:", long_msg)
wxflow.executable.ProcessError: Command exited with status 137:
'mpiexec' '-l' '-n' '120' '-ppn' '120' '--cpu-bind' 'depth' '--depth' '1' '/u/wen.meng/ptmp/gw_test/RUNDIR/upp.231494/upp.x'
^[[38;5;196m2024-03-18 17:12:57,639 - ERROR    - upp         : FATAL ERROR: Error occurred during execution of mpiexec -l -n 120 -ppn 120 --cpu-bind depth --depth 1 /u/wen.meng/ptmp/gw_test/RUNDIR/upp.231494/upp.x^[
```

### Additional information

The log file 'gfs_upp.o121904081-failture' can be found at /u/wen.meng/noscrub/ncep_post/gw_v17/driver on Cactus.

### Do you have a proposed solution?

Yes. It appears all tasks allocated in one node cause the failure on WCOSS2 as:
Executing mpiexec -l -n 120 -ppn 120 --cpu-bind depth --depth 1 /u/wen.meng/ptmp/gw_test/RUNDIR/upp.231494/upp.x

My proposed solution would be in global-workflow/parm/config/gfs/config.resources:
change
   export npe_node_upp=${npe_upp}
into 
   export npe_node_upp=40

In that way, 120 tasks will be allocated in 3 or more nodes.

### PR DESCRIPTION
**Pull request by aerorahul**
# Description
This PR:
- adds the option of running the ufs-weather-model with traditional threading in addition to ESMF-managed threading. See new toggle `USE_ESMF_THREADING=YES|NO` set in `config.fcst`
- does not change the current default of using ESMF-managed threading.  Traditional threading use might need a little more fine tuning for the job-card specification.  This will be achieved when the UFSWM RT completely switches over to traditional threading
- updates the hash of the ufs-weather-model to the PR https://github.com/ufs-community/ufs-weather-model/pull/2172

Resolves #2277

In addition to the above stated objectives, this PR also addresses open issues.  In particular, this PR:
- adds a newline at the end of `diag_table_aod`.  Fixes #2407 @zhanglikate 
- reserves more memory on WCOSS2 for offline UPP when running at C768.  Fixes #2408 @WenMeng-NOAA 

# Type of change
- New feature (adds functionality)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
- Tested on WCOSS2 with `C384_S2SWA` test

# Checklist
- [ ] Any dependent changes have been merged and published
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] I have made corresponding changes to the documentation if necessary

### CODE REVIEW

File: parm/config/gefs/config.fcst [hunk 1]
```diff
@@ -5,6 +5,8 @@
 
 echo "BEGIN: config.fcst"
 
+export USE_ESMF_THREADING="YES"  # Toggle to use ESMF-managed threading or traditional threading in UFSWM
```
Comment (WalterKolczynski-NOAA) [inline parm/config/gefs/config.fcst:8]: Do we want to template it and set it in the hosts files so it is machine-dependent?

File: parm/config/gefs/config.ufs [hunk 2]
```diff
@@ -433,27 +447,30 @@ fi
 
 # Set the name of the UFS (previously nems) configure template to use
 # Default ufs.configure templates for supported model configurations
+if [[ "${USE_ESMF_THREADING:-}" == "YES" ]]; then
+  tmpl_suffix="_esmf"
+fi
 case "${model_list}" in
   atm)
-    default_template="${PARMgfs}/ufs/ufs.configure.atm.IN"
+    default_template="${HOMEgfs}/parm/ufs/ufs.configure.atm${tmpl_suffix:-}.IN"
```
Comment (WalterKolczynski-NOAA) [inline parm/config/gefs/config.ufs]: Shouldn't all these retain `${PARMgfs}` over `${HOMEgfs}/parm`?

**Review by WalterKolczynski-NOAA — DISMISSED** [review]

File: parm/config/gefs/config.fcst (hunk 1 above)
Comment (aerorahul) [reply parm/config/gefs/config.fcst:8]: It is unclear yet if this will/should be machine dependent.

File: parm/config/gefs/config.ufs (hunk 2 above)
Comment (aerorahul) [reply parm/config/gefs/config.ufs]: yes. done in e7e2b9a3

**Comment by aerorahul** [comment]
UFS-weather-model [PR 2172](https://github.com/ufs-community/ufs-weather-model/pull/2172) was merged.  This PR is now ready for sync and review.

File: sorc/link_workflow.sh [hunk 3]
```diff
@@ -166,12 +166,18 @@ declare -a ufs_templates=("model_configure.IN" \
...
+                          "ufs.configure.s2s.IN" \
                           "ufs.configure.s2s_esmf.IN" \
+                          "ufs.configure.s2sa.IN" \
                           "ufs.configure.s2sa_esmf.IN" \
+                          "ufs.configure.s2sw.IN" \
                           "ufs.configure.s2sw_esmf.IN" \
-                          "ufs.configure.s2swa_esmf.IN" )
+                          "ufs.configure.s2swa.IN" \
+                          "ufs.configure.s2swa_esmf.IN" \
+                          "ufs.configure.leapfrog_atm_wav.IN" )
```
Comment (DavidHuber-NOAA) [inline sorc/link_workflow.sh]: `ufs.configure.leapfrog_atm_wav_esmf.IN` also needs to be linked in.
```suggestion
                          "ufs.configure.leapfrog_atm_wav.IN" \
                          "ufs.configure.leapfrog_atm_wav_esmf.IN" )
```

Comment (aerorahul) [reply sorc/link_workflow.sh]: right. thanks for catching that.

**Review by DavidHuber-NOAA — APPROVED** [review]
Looks good to me!

File: parm/config/gfs/config.resources [hunk 4]
```diff
@@ -692,9 +706,13 @@ case ${step} in
...
       ;;
-      "C192" | "C384" | "C768")
+      "C192" | "C384")
+        export npe_upp=120
+        export memory_upp="96GB"
+      ;;
+      "C768")
         export npe_upp=120
-        export memory_upp="48GB"
+        if [[ ${machine} == "WCOSS2" ]]; then export memory_upp="480GB" ; fi
```
Comment (KateFriedman-NOAA) [inline parm/config/gfs/config.resources:716]: What about the memory value for the other machines for C768? Run upp exclusively elsewhere?

I don't see memory for the lowest resolutions, do we need that too or also run exclusively?

**Review by KateFriedman-NOAA — APPROVED** [review]
Looks good, left one comment to consider but approve pending any CI testing. Thanks @aerorahul !

File: parm/config/gfs/config.resources (hunk 4 above)
Comment (aerorahul) [inline parm/config/gfs/config.resources:716]:
```suggestion
        export memory_upp="96GB"
        if [[ ${machine} == "WCOSS2" ]]; then export memory_upp="480GB" ; fi
```

Comment (aerorahul) [reply parm/config/gfs/config.resources:716]: how about now.  This should atleast cover what was there before.

Comment (KateFriedman-NOAA) [reply parm/config/gfs/config.resources:716]: Works for me, thanks @aerorahul !

**Comment by aerorahul** [comment]
> Experiment C96C48_hybatmDA failed on Orion
> 
> Error logs:
> 
> RUNTESTS/COMROOT/C48_S2SWA_gefs_65830b8b/logs/2021032312/ice_prod_mem001_f006.log

The error file `/work2/noaa/stmp/CI/ORION/2384/RUNTESTS/COMROOT/C48_S2SWA_gefs_65830b8b/logs/2021032312/ice_prod_mem001_f006.log` says:
```
$> cat /work2/noaa/stmp/CI/ORION/2384/RUNTESTS/COMROOT/C48_S2SWA_gefs_65830b8b/logs/2021032312/ice_prod_mem001_f006.log
Begin oceanice_products.sh at Thu Mar 21 15:33:12 UTC 2024
+ oceanice_products.sh[11]: . /work2/noaa/stmp/CI/ORION/2384/gefs/ush/load_fv3gfs_modules.sh
++ load_fv3gfs_modules.sh[4]: [[ NO == \N\O ]]
++ load_fv3gfs_modules.sh[5]: echo 'Loading modules quietly...'
Loading modules quietly...
++ load_fv3gfs_modules.sh[6]: set +x
Resetting modules to system default. Reseting $MODULEPATH back to system default. All extra directories will be removed from $MODULEPATH.
Lmod Warning: MODULEPATH is undefined.



Lmod has detected the following error: The following module(s) are unknown:
"contrib"

Please check the spelling or version number. Also try "module spider ..."
It is also possible your cache file is out-of-date; it may help to try:
  $ module --ignore-cache load "contrib"

Also make sure that all modulefiles written in TCL start with the string
#%Module



End oceanice_products.sh at 15:33:12 with error code 1 (time elapsed: 00:00:00)
```

I am not sure why this particular job had issues with finding `contrib`.  This should have been the same case for `stage_ic.sh` and `fcst.sh`.

**Comment by aerorahul** [comment]
> Experiment C96C48_hybatmDA failed on Orion
> 
> Error logs:
> 
> RUNTESTS/COMROOT/C48_S2SWA_gefs_65830b8b/logs/2021032312/ice_prod_mem001_f006.log

@TerrenceMcGuinness-NOAA 
The experiment fail log is not for the experiment that failed.

**Comment by aerorahul** [comment]
Looking at the log for another failed job `eobs`, it also failed with the same error on the first attempt:
```
$> cat /work2/noaa/stmp/CI/ORION/2384/RUNTESTS/COMROOT/C96C48_hybatmDA_65830b8b/logs/2021122100/enkfgdaseobs.log.0
Begin eobs.sh at Thu Mar 21 15:33:11 UTC 2024
+ eobs.sh[7]: . /work2/noaa/stmp/CI/ORION/2384/gfs/ush/load_fv3gfs_modules.sh
++ load_fv3gfs_modules.sh[4]: [[ NO == \N\O ]]
++ load_fv3gfs_modules.sh[5]: echo 'Loading modules quietly...'
Loading modules quietly...
++ load_fv3gfs_modules.sh[6]: set +x
Resetting modules to system default. Reseting $MODULEPATH back to system default. All extra directories will be removed from $MODULEPATH.
Lmod Warning: MODULEPATH is undefined.



Lmod has detected the following error: The following module(s) are unknown:
"contrib"

Please check the spelling or version number. Also try "module spider ..."
It is also possible your cache file is out-of-date; it may help to try:
  $ module --ignore-cache load "contrib"

Also make sure that all modulefiles written in TCL start with the string
#%Module



End eobs.sh at 15:33:11 with error code 1 (time elapsed: 00:00:00)
```
This indicates that the issue is with system being unstable or something unrelated to the changes in this PR.

**Comment by DavidHuber-NOAA** [comment]
The failed jobs all occurred on the Orion-16-06 compute node:
```
   C48_S2SWA_gefs_65830b8b_ice_prod_mem001_f006_12     FAILED 17184639              Orion-16-06 
   C48_S2SWA_gefs_65830b8b_ice_prod_mem001_f006_12     FAILED 17184711              Orion-16-06 
          C96C48_hybatmDA_65830b8b_enkfgdaseobs_00     FAILED 17184695         Orion-16-[06-07]
```

I suspect this is some sort of node lustre/communication issue.  I'll open a ticket with RDHPCS.

**Comment by TerrenceMcGuinness-NOAA** [comment]
[gdasanal.log](https://jenkins.epic.oarcloud.noaa.gov/job/global-workflow/job/EMC-Global-Pipeline/job/PR-2384/6/artifact/RUNTESTS/COMROOT/C96_atm3DVar_668ede00/logs/2021122100/gdasanal.log) **Segfaulted**

/work2/noaa/stmp/CI/ORION/2384/RUNTESTS/COMROOT/C96C48_hybatmDA_668ede00/logs/2021122100/gfsanal.log

**Comment by aerorahul** [comment]
> /work2/noaa/stmp/CI/ORION/2384/RUNTESTS/COMROOT/C96C48_hybatmDA_668ede00/logs/2021122100/gfsanal.log

It seems to have segfaulted immediately:
```
+ exglobal_atmos_analysis.sh[912]: srun -l --export=ALL -n 84 --cpus-per-task=5 /work/noaa/stmp/mterry/RUNDIRS/C96C48_hybatmDA_668ede00/anal.138549/gsi.x
srun: warning: can't honor --ntasks-per-node set to 8 which doesn't match the requested tasks 84 with the number of requested nodes 11. Ignoring --ntasks-per-node.
33: [Orion-20-02:100519:0:100519]     address.c:1052 Assertion `*addr_version == UCP_OBJECT_VERSION_V2' failed: addr version 6
36: [Orion-20-02:100522:0:100522]     address.c:1052 Assertion `*addr_version == UCP_OBJECT_VERSION_V2' failed: addr version 6
70: [Orion-25-40:262060:0:262060] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x2ae119ed0008)
71: [Orion-25-40:262061:0:262061] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x2ba29677c008)
72: [Orion-25-40:262062:0:262062] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x2b5eec527008)
73: [Orion-25-40:262063:0:262063] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x2aea0cae3008)
74: [Orion-25-40:262064:0:262064] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x2b9208977008)
75: [Orion-25-40:262065:0:262065] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x2ad76baf3008)
76: [Orion-25-40:262066:0:262066] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x2b94bc56d008)
```

The first attempt had a similar issue:
```
+ exglobal_atmos_analysis.sh[912]: srun -l --export=ALL -n 84 --cpus-per-task=5 /work/noaa/stmp/mterry/RUNDIRS/C96C48_hybatmDA_668ede00/anal.134326/gsi.x
srun: warning: can't honor --ntasks-per-node set to 8 which doesn't match the requested tasks 84 with the number of requested nodes 11. Ignoring --ntasks-per-node.
70: [1711046630.238503] [Orion-25-40:260437:0]         select.c:438  UCX  ERROR no active messages transport to <no debug data>: self/memory - Destination is unreachable, rc/mlx5_0:1 - Destination is unreachable, rc_mlx5/mlx5_0:1 - Destination is unreachable, rdmacm/sockaddr - no am bcopy
71: [1711046630.238494] [Orion-25-40:260438:0]         select.c:438  UCX  ERROR no active messages transport to <no debug data>: self/memory - Destination is unreachable, rc/mlx5_0:1 - Destination is unreachable, rc_mlx5/mlx5_0:1 - Destination is unreachable, rdmacm/sockaddr - no am bcopy
72: [1711046630.238503] [Orion-25-40:260439:0]         select.c:438  UCX  ERROR no active messages transport to <no debug data>: self/memory - Destination is unreachable, rc/mlx5_0:1 - Destination is unreachable, rc_mlx5/mlx5_0:1 - Destination is unreach
able, rdmacm/sockaddr - no am bcopy
```

I don't think Orion is feeling well today.

**Comment by TerrenceMcGuinness-NOAA** [comment]
"Hand" moving Label to **CI-Orion-Fail** due to immediate seg fault

**Review by WalterKolczynski-NOAA — APPROVED** [review]
Conditionally approved pending successful completion of CI tests.

**Pull request merged**